src/docfx/docset/OutputProvider.cs

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+
+namespace Microsoft.Docs.Build
+{
+    internal class OutputProvider
+    {
+        private readonly ConcurrentDictionary<string, Lazy<string>> _outputMapping = new ConcurrentDictionary<string, Lazy<string>>();
+
+        public string GetOutputPath(Document file)
+            => _outputMapping.GetOrAdd(file.FilePath, new Lazy<string>(() =>
+            {
+                var monikers = file.Docset.MonikersProvider.GetMonikers(file);
+                if (monikers.Count == 0)
+                {
+                    return file.SitePath;
+                }
+                return Path.Combine(MonikerUtility.GetMonikersHash(monikers), file.SitePath);
```

Comment:
>monikers [](start = 67, length = 8)

Is this order stable, even across products?

Reply:
The result of `MonikerDefinitionUrl` api call should be order stable across products, right?

---
In reply to: [232146903](https://github.com/dotnet/docfx/pull/3596#discussion_r232146903) [](ancestors = 232146903)

Reply:
good catch！ we should make it stable

---
In reply to: [232154947](https://github.com/dotnet/docfx/pull/3596#discussion_r232154947) [](ancestors = 232154947,232146903)

Reply:
add a test for it?